### ISSUE TITLE
cannot test  an Action that handles a promise

### ISSUE THREAD

**Issue by soundobj** [goatslacker/alt#409]
Hi there I Have an Action that handles a promise and I am having problems testing it

the action has a private method handleResponse() to load data asynchronously it also has a Service injected via a setService() action.

 var handleResponse = function(promise) {
        submissionsActions.onLoadSubmissions({
            processing: true,
            loading: false
        });

        // Only show the loading message if the response takes more than half a second
        var loadingTimer = setTimeout(function() {
            submissionsActions.onLoadSubmissions({
                processing: true,
                loading: true
            });
        }.bind(this), 500);

        promise
            .then(function(submissions) {
                submissionsActions.initStore(submissions);
                submissionsActions.onLoadSubmissionsSuccess(submissions);
            }.bind(this))
            ["finally"](function() {
                clearInterval(loadingTimer);
            })
            ["catch"](function(error) {
                try {
                    submissionsActions.onLoadSubmissionsFail(error);
                } catch (ex) {
                    console.error(ex);
                }
            }.bind(this));
    };

var submissionsActions = alt.createActions({
        loadSubmissions: function() {
            handleResponse.call(this, service.getSubmissions());
        },
        setService: function(ser) {
            console.log('Submissions Actions setting service',ser);
            service = ser;
        },
    });

On my test I am using Sinon to mock the service and its response to getSubmissions:

service = {
                getSubmissions: function() {
                    console.log('IWAS CALLED');
                    return {"submissions": []};
                },
                refreshSubmissions: function() {}
            };

            Actions.dispatch = function() {};
            Actions.setService(service);

                defer = Q.defer();
                defer.resolve({"submissions": []});
                // @TODO: service is now
                //sinon.stub(service, 'getSubmissions').returns(defer.promise);
                Actions.loadSubmissions();

                // Push Q
                this.clock.tick();

when the promise is fulfilled: i.e
 .then(function(submissions) {

I am getting the below error:
Uncaught ReferenceError: submissions is not defined
    at AltAction.eval (eval at evaluate (unknown source), <anonymous>:1:1)
    at Object.InjectedScript._evaluateOn (<anonymous>:895:55)
    at Object.InjectedScript._evaluateAndWrap (<anonymous>:828:34)
    at Object.InjectedScript.evaluateOnCallFrame (<anonymous>:954:21)
    at AltAction.handleResponse 

It seems that AltAction.eval() is ignoring the resolved promise. 

Any help would be greatly appreciated.

**Comment by goatslacker** [comment]
hard to follow the code but I'm thinking you need to mock the promise value that gets resolved.

**Comment by Emibob** [comment]
Hey there, we got to the part where we resolve the Promise. What we've noticed is we can either spy on the `alt.actions.NS.action` but the `this.actions.action` inside the promise references the actual function. Conversely, if we spy on the ActionClass.action then `this.actions.action
Implementation:
```
'use strict';

const alt = require('../alt');
const FetchApi = require('../util/FetchApi');

export class LandingPageActions {

  updateEdition(date) {
    this.dispatch(date);
  }

  fetchFailed(errorMessage) {
  	this.dispatch(errorMessage);
  }

  fetchData(path) {
    
  	this.dispatch();
  	var path = path || 'current?state=weekend';
  	FetchApi.fetch(path)
  		.then((data) => {
  		  this.actions.updateEdition(data);
  		})
  		.catch((e) => {
  			console.log('Response Error', e);
  			this.actions.fetchFailed(e);
  		});
  }
}

module.exports = alt.createActions(LandingPageActions);
```
unit test:
```
jest.dontMock('../LandingPageActions');

const LandingPageActions = require('../LandingPageActions');
const Actions = require('../../alt').actions;
const Dispatcher = require('../../alt').dispatcher;
const FetchApi = require('../../util/FetchApi');


Dispatcher.dispatch = jest.genMockFunction();

describe('LandingPage Actions', function() {
  describe('#updateEdition', function() {
    beforeEach(function() {
    	this.date = 'Aug 4, 2015 11:30 AM';
      LandingPageActions.updateEdition(this.date);
    });

    it('calls dispatch with the non-transformed date', function() {
      expect(Dispatcher.dispatch.mock.calls[0][0].data).toEqual(this.date);
    });
  });

  describe('#fetchData', function() {
  	beforeEach(function() {
  		this.response = {publish_date: 'Aug 8, 2015 11:30 AM'};
  		Actions.LandingPageActions.updateEdition = jest.genMockFunction();
  		FetchApi.fetch = jest.genMockFunction().mockImplementation(function(path) {
  			return Promise.resolve(this.response);
  		}.bind(this));
  	});

    describe('when called without arguments', function() {
   		it('calls fetch with the correct default path', function() {
 
   	  	LandingPageActions.fetchData();
   	  	expect(LandingPageActions.updateEdition).toBeCalledWith(this.response);
   	 	});
   	});

  });
});
```

**Comment by goatslacker** [comment]
Hmm that's a bit tricky, why not just spy on the dispatcher and check the action type there? Spying on the dispatcher is usually the best way to test actions flowing through.
